File: app/server/notify/channels/gotify.ts
/**
 * Gotify 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'gotify' });

export const gotifyChannel: NotifyChannel = {
    name: 'gotify',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('GOTIFY_URL', 'GOTIFY_TOKEN')) {
                resolve({ success: false, message: 'Gotify 配置不完整' });
                return;
            }

            const GOTIFY_URL = getConfig('GOTIFY_URL');
            const GOTIFY_TOKEN = getConfig('GOTIFY_TOKEN');
            const GOTIFY_PRIORITY = getConfig('GOTIFY_PRIORITY') || 0;

            const url = `${GOTIFY_URL}/message?token=${GOTIFY_TOKEN}`;
            const body = `title=${encodeURIComponent(text)}&message=${encodeURIComponent(desp)}&priority=${GOTIFY_PRIORITY}`;

            $.post({
                url,
                body,
                headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
            }, (err, _resp, data) => {
                if (err) {
                    logger.error({ err }, 'Gotify 发送失败');
                    resolve({ success: false, error: err as Error });
                } else {
                    const result = data as { id?: number; message?: string };
                    if (result?.id) {
                        logger.info('Gotify 发送成功');
                        resolve({ success: true, message: '发送成功' });
                    } else {
                        logger.warn({ body: data }, 'Gotify 发送失败');
                        resolve({ success: false, message: result?.message || '发送失败' });
                    }
                }
            });
        });
    }
};

if (hasConfig('GOTIFY_URL', 'GOTIFY_TOKEN')) {
    gotifyChannel.enabled = true;
}

export default gotifyChannel;

File: app/server/internal/notify/channels/gotify.go
package channels

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"net/url"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type Gotify struct {
	enabled bool
}

func (c *Gotify) Name() string            { return "gotify" }
func (c *Gotify) Enabled() bool           { return c.enabled }
func (c *Gotify) SetEnabled(enabled bool) { c.enabled = enabled }

func (c *Gotify) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("GOTIFY_URL", "GOTIFY_TOKEN") {
		return notify.NotifyResult{Success: false, Message: "Gotify 配置不完整"}
	}

	gotifyURL := notify.GetConfig("GOTIFY_URL")
	gotifyToken := notify.GetConfig("GOTIFY_TOKEN")
	gotifyPriority := notify.GetConfig("GOTIFY_PRIORITY")
	if gotifyPriority == "" {
		gotifyPriority = "0"
	}

	apiURL := fmt.Sprintf("%s/message?token=%s", gotifyURL, gotifyToken)

	body := fmt.Sprintf("title=%s&message=%s&priority=%s",
		url.QueryEscape(text), url.QueryEscape(desp), gotifyPriority)

	resp, err := notify.HTTPPost(apiURL, notify.HttpRequestOptions{
		Body:    body,
		Headers: map[string]string{"Content-Type": "application/x-www-form-urlencoded"},
	})
	if err != nil {
		slog.Error("Gotify 发送失败", "err", err)
		return notify.NotifyResult{Success: false, Message: fmt.Sprintf("发送失败: %v", err), Err: err}
	}

	var result struct {
		ID      float64 `json:"id"`
		Message string  `json:"message"`
	}
	if err := json.Unmarshal([]byte(resp.Body), &result); err != nil {
		slog.Error("Gotify 解析响应失败", "err", err)
		return notify.NotifyResult{Success: false, Message: "解析响应失败", Err: err}
	}

	if result.ID != 0 {
		slog.Info("Gotify 发送成功")
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	msg := result.Message
	if msg == "" {
		msg = "发送失败"
	}
	slog.Warn("Gotify 发送失败", "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: msg}
}

func init() {
	ch := &Gotify{}
	if notify.HasConfig("GOTIFY_URL", "GOTIFY_TOKEN") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
